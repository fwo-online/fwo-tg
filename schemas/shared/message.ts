import type { CharacterPublic } from '@/character/characterPublic';
import type { Character } from '@/character';
import type { Player } from './player';
import type { GameStatus } from '@/game';
import type { Action } from './action';
import type { RPC } from './rpc';

export type ClientToServerMessage = Message<{
  character: [callback: (character: Character) => void];
  'lobby:enter': [callback: (characters: CharacterPublic[]) => void];
  'lobby:leave': [];
  'lobby:start': [];
  'lobby:stop': [];
  'game:connected': [callback: (payload: RPC<{ players: Record<string, Player> }>) => void];
  'game:order': [
    order: {
      power: number;
      action: string;
      target: string;
    },
    callback: (
      payload: RPC<{
        actions: Action[];
        magics: Action[];
        skills: Action[];
        power: number;
        orders: {
          power: number;
          action: string;
          target: string;
        }[];
      }>,
    ) => void,
  ];
}>;

type Message<T extends Record<string, unknown[]>> = {
  [K in keyof T]: (...args: T[K]) => void;
};

export type ServerToClientMessage = Message<{
  character: [character: Character];
  'lobby:enter': [character: CharacterPublic];
  'lobby:leave': [character: CharacterPublic];
  'lobby:list': [characters: CharacterPublic[]];
  'lobby:start': [character: CharacterPublic];
  'lobby:stop': [character: CharacterPublic];
  'game:start': [gameID: string];
  'game:end': [];
  'game:startOrders': [actions: { actions: Action[]; magics: Action[]; skills: Action[] }];
  'game:endOrders': [];
  'game:startRound': [
    {
      round: number;
      status: Partial<Record<string, GameStatus[]>>;
    },
  ];
  'game:endRound': [{ dead: Player[] }];
  'game:kick': [{ reason: string; player: Player }];
  'game:preKick': [{ reason: string; player: Player }];
}>;
