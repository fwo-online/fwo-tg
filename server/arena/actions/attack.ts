import type { Affect } from '../Constuructors/interfaces/Affect';
import PhysConstructor from '../Constuructors/PhysConstructor';
import CastError from '../errors/CastError';
/**
 * Физическая атака
 */
class Attack extends PhysConstructor implements Affect {
  constructor() {
    super({
      name: 'attack',
      displayName: 'Атака',
      desc: 'Атака',
      lvl: 0,
      orderType: 'all',
    });
  }

  /**
 * Кастомный обработчик атаки
 */
  run() {
    if (!this.params) {
      return;
    }
    const { initiator, target } = this.params;

    target.flags.isHited = {
      initiator: initiator.nick, val: this.status.effect,
    };

    target.stats.down('hp', this.status.effect);
  }

  preAffect: Affect['preAffect'] = ({ params: { initiator, target, game } }): undefined => {
    if (target.flags.isHited) {
      throw new CastError(this.getSuccessResult({ initiator: target, target: initiator, game }));
    }
  };
}
export default new Attack();
