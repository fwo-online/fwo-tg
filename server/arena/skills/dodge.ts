import { bold, italic } from '@/utils/formatString';
import type { Affect } from '../Constuructors/interfaces/Affect';
import { Skill } from '../Constuructors/SkillConstructor';
import type { ActionType, SuccessArgs } from '../Constuructors/types';
import CastError from '../errors/CastError';
import MiscService from '../MiscService';

const dodgeableWeaponTypes = [
  'thrust', // –∫–æ–ª—é—â–µ–µ
  'cut', // —Ä–µ–∂—É—â–µ–µ
  'chop', // —Ä—É–±—è—â–µ–µ
  'range', // –º–µ—Ç–∞—Ç–µ–ª—å–Ω–æ–µ
  'stun', // –æ–≥–ª—É—à–∞—é—â–µ–µ
];

/**
 * –£–≤–µ—Ä—Ç–∫–∞
 */
class Dodge extends Skill implements Affect {
  actionType: ActionType = 'dodge';

  constructor() {
    super({
      name: 'dodge',
      displayName: 'üêç –£–≤–µ—Ä—Ç–∫–∞',
      desc: '–®–∞–Ω—Å —É–≤–µ—Ä–Ω—É—Ç—å—Å—è –æ—Ç –æ–¥–Ω–æ–π –∏–ª–∏ –Ω–µ—Å–∫–æ–ª—å–∫–∏—Ö –∞—Ç–∞–∫(—Ç–æ–ª—å–∫–æ –ø—Ä–æ—Ç–∏–≤ –∫–æ–ª—é—â–µ–≥–æ, —Ä–µ–∂—É—â–µ–≥–æ, —Ä—É–±—è—â–µ–≥–æ, –º–µ—Ç–∞—Ç–µ–ª—å–Ω–æ–≥–æ, –æ–≥–ª—É—à–∞—é—â–µ–≥–æ –æ—Ä—É–∂–∏—è)',
      cost: [10, 12, 14, 16, 18, 20],
      proc: 20,
      baseExp: 50,
      costType: 'en',
      orderType: 'self',
      aoeType: 'target',
      chance: [75, 80, 85, 90, 95, 99],
      effect: [1.2, 1.25, 1.3, 1.35, 1.4, 1.45],
      profList: { l: 2 },
      bonusCost: [10, 20, 30, 40, 60, 80],
    });
  }

  run() {
    const { initiator } = this.params;
    const initiatorSkillLvl = initiator.skills[this.name];
    initiator.flags.isDodging = this.effect[initiatorSkillLvl - 1] * initiator.stats.val('attributes.dex');
  }

  preAffect: Affect['preAffect'] = ({ params: { initiator, target, game } }): undefined => {
    const isDodgeable = initiator.weapon.isOfType(dodgeableWeaponTypes);

    if (target.flags.isDodging && isDodgeable) {
      const initiatorDex = initiator.stats.val('attributes.dex');
      const dodgeFactor = Math.round(target.flags.isDodging / initiatorDex);
      const chance = Math.round(Math.sqrt(dodgeFactor) + (10 * dodgeFactor) + 5);

      if (chance > MiscService.rndm('1d100')) {
        this.calculateExp();

        throw new CastError(this.getSuccessResult({ initiator: target, target: initiator, game }));
      }
    }
  };

  customMessage(args: SuccessArgs) {
    return `${bold(args.initiator.nick)} –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª ${italic(this.displayName)}`;
  }
}

export default new Dodge();
