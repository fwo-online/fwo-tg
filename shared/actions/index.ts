import type { CharacterClass } from '@/character';

export type Magic = {
  name: string;
  description: string;
  displayName: string;
  effect: string[];
  cost: number;
  costType: CostType;
  lvl: number;
  orderType: OrderType;
  effectType?: EffectType;
};

export type Skill = {
  name: string;
  displayName: string;
  description: string;
  classList: Partial<Record<CharacterClass, number>>;
  bonusCost: number[];
  orderType: OrderType;
  cost: number;
  costType: CostType;
};

export enum CostType {
  MP = 'mp',
  EN = 'en',
}

export enum OrderType {
  All = 'all',
  Any = 'any',
  Enemy = 'enemy',
  Self = 'self',
  Team = 'team',
  TeamExceptSelf = 'teamExceptSelf',
}

export enum EffectType {
  Acid = 'acid',
  Fire = 'fire',
  Lighting = 'lighting',
  Frost = 'frost',
  Physical = 'physical',
  Clear = 'clear',
}

export type ActionType =
  | 'magic'
  | 'dmg-magic'
  | 'dmg-magic-long'
  | 'aoe-dmg-magic'
  | 'magic-long'
  | 'skill'
  | 'phys'
  | 'heal-magic'
  | 'heal'
  | 'protect'
  | 'dodge'
  | 'passive'
  | 'regeneration';

export type ReservedFailReason =
  | 'NO_INITIATOR'
  | 'NO_TARGET'
  | 'NO_MANA'
  | 'NO_ENERGY'
  | 'CHANCE_FAIL'
  | 'GOD_FAIL'
  | 'HEAL_FAIL'
  | 'SKILL_FAIL'
  | 'PHYS_FAIL'
  | 'NO_WEAPON';
