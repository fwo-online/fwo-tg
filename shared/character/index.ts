import type { CharacterAttributes } from './characterAttributesSchema';
import { CharacterClass } from './characterClassSchema';
import type { Attributes } from '@/shared/attributes';

export const characterAttributesByClass: Record<CharacterClass, CharacterAttributes> = {
  [CharacterClass.Warrior]: {
    str: 10,
    dex: 8,
    int: 3,
    wis: 3,
    con: 6,
  },
  [CharacterClass.Archer]: {
    str: 3,
    dex: 10,
    int: 8,
    wis: 3,
    con: 6,
  },
  [CharacterClass.Mage]: {
    str: 3,
    dex: 3,
    int: 8,
    wis: 10,
    con: 6,
  },
  [CharacterClass.Priest]: {
    str: 3,
    dex: 3,
    int: 10,
    wis: 8,
    con: 6,
  },
};

export type CharacterDynamicAttributes = Attributes & {
  maxTarget: number;
  spellLength: number;
};

export * from './characterAttributesSchema';
export * from './characterClassSchema';
export * from './characterSchema';
export * from './createCharacterSchema';
export * from './characterPublic';
