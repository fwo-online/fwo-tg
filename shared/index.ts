export * from './character';
export * from './inventory';
export * from './item';
export * from './itemSet';
export * from './shared';
export * from './actions';
export * from './game';
export * from './clan';
