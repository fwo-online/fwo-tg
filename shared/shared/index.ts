export * from './id';
export * from './modifiers';
export * from './player';
export * from './orderSchema';
export * from './message';
export * from './attributes';
export * from './action';
