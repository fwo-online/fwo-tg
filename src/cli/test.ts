import { $ } from 'bun';

async function test() {
  try {
    await $`docker-compose up -d --build db`;

    await $`bun test --preload ./test.setup.ts`;

    await $`docker-compose stop db`;
    process.exit(0);
  } catch {
    process.exit(1);
  }
}

void test();
