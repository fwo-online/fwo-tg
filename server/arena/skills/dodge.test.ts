import { afterEach, beforeEach, describe, expect, it } from 'bun:test';
import { CharacterClass } from '@fwo/shared';
import type GameService from '@/arena/GameService';
import TestUtils from '@/utils/testUtils';
import attack from '../actions/attack';
import dodge from './dodge';

// npm t server/arena/skills/dodge.test.ts

describe('dodge', () => {
  let game: GameService;

  beforeEach(async () => {
    attack.registerPreAffects([dodge]);

    game = await TestUtils.createGame([
      {
        weapon: { type: 'chop' },
      },
      {
        prof: CharacterClass.Archer,
        skills: { dodge: 3 },
      },
    ]);

    TestUtils.mockRandom();
  });

  afterEach(() => {
    TestUtils.restoreRandom();
  });

  it('target should dodge attack if has initiator more dex', async () => {
    game.players.players[0].proc = 1;
    game.players.players[1].stats.set('attributes.dex', 9999);

    dodge.cast(game.players.players[1], game.players.players[1], game);
    attack.cast(game.players.players[0], game.players.players[1], game);

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('target should not dodge attack if target has more dex', async () => {
    game.players.players[0].proc = 1;
    game.players.players[0].stats.set('attributes.dex', 9999);

    dodge.cast(game.players.players[1], game.players.players[1], game);
    attack.cast(game.players.players[0], game.players.players[1], game);

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
