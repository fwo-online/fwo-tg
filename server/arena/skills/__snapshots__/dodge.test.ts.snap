// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`dodge target should dodge attack if has initiator more dex 1`] = `
"*alias* использовал _🐍 Увертка_

*asperiores* пытался атаковать *alias*, но у него не получилось
_🐍 Увертка_ *alias*: 📖50"
`;

exports[`dodge target should not dodge attack if target has more dex 1`] = `
"*explicabo* использовал _🐍 Увертка_

*repellat* рубанул *explicabo* _Эпическим Мечом_ и нанёс *4.85* урона
\\[ explicabo 💥 💔-4.85/3.15 📖39 ]"
`;
